### PR TITLE
RevEng: Allow user to specify context class name

### LINKED ISSUE #3047: Feature Request: Scaffolding a database should allow to specify the context name
Currently (beta7) when scaffolding a database the derived DbContext class will pick the database name suffixed with the "Context" word.

For example:
```
dnx ef dbcontext scaffold "Server=myserver;Database=Foo;user id=me;password=pwd" EntityFramework.SqlServer
```

Produces a class named `FooContext`. The command should allow passing the context name as a parameter:
```
dnx ef dbcontext scaffold "Server=myserver;Database=Foo;user id=me;password=pwd" EntityFramework.SqlServer --context-name MyDbContext
```

### PR DESCRIPTION
**Pull request by lajones**
Fixes #3047.

### CODE REVIEW

**Comment by natemcmaster** [comment]
Should we also check if the context name is a valid C# identifier?

**Comment by lajones** [comment]
@natemcmaster I figured the result would be a context class that wouldn't compile and they'd learn that way. But since we've both considered it perhaps I should just fail early instead. I'll update.

**Comment by lajones** [comment]
@natemcmaster all fixed - please review latest commit.

File: src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs [hunk 1]
```diff
@@ -207,6 +207,32 @@ public virtual string GenerateLiteral([NotNull] object value)
             return string.Format(CultureInfo.InvariantCulture, "{0}", value);
         }
 
+        public virtual bool IsValidCSharpIdentifier([NotNull] string identifier)
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs]: Why not use `Microsoft.CodeAnalysis.CSharp.SyntaxFacts.IsValidIdentifier`?

File: src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs [hunk 2]
```diff
@@ -207,6 +207,32 @@ public virtual string GenerateLiteral([NotNull] object value)
...
+            }
+
+            if (_cSharpKeywords.Contains(identifier))
+            {
+                return false;
+            }
+
+            if (_invalidCharsRegex.IsMatch(identifier))
+            {
+                System.IO.File.AppendAllText(@"C:\work\150924_ContextClassName", "Identifier, " + identifier + ", matches _invalidCharsRegex" + Environment.NewLine);
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs]: Debug?

**Comment by natemcmaster** [comment]
:shipit: (with the mentioned above)

File: src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs]: Oops. Will fix.

File: src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs (hunk 1 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Utilities/CSharpUtilities.cs]: Good point. Will use that instead.

File: src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringConfiguration.cs [hunk 3]
```diff
@@ -29,6 +30,13 @@ public virtual void CheckValidity()
                 throw new ArgumentException(Strings.ProjectPathRequired);
             }
 
+            if (!string.IsNullOrWhiteSpace(ContextClassName)
+                && !CSharpUtilities.Instance.IsValidCSharpIdentifier(ContextClassName))
```
Comment (divega) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringConfiguration.cs]: I am personally not sure we need to do this, but FWIW, the namespace should be a valid identifier as well.

Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringConfiguration.cs]: @divega The namespace is constructed from the project and sub-directory names and already goes through the 'making sure it's a valid identifier' process.

Comment (divega) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringConfiguration.cs]: OK, thanks!

**Comment by lajones** [comment]
Checked in with commit 4f133c1.

**Pull request closed without merging**